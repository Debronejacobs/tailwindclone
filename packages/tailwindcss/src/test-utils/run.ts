import { optimizeCss } from '@tailwindcss/optimize'
import { compile } from '..'

export function compileCss(css: string, candidates: string[] = []) {
  return optimizeCss(compile(css).build(candidates)).trim()
}

export function run(candidates: string[]) {
  return optimizeCss(compile('@tailwind utilities;').build(candidates)).trim()
}
