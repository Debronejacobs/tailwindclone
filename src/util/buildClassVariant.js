import escapeClassName from './escapeClassName'

export default function buildClassVariant(className, variantName, variantsAsSuffix, separator) {
  if (variantsAsSuffix) {
    return `${className}${escapeClassName(separator)}${variantName}`
  }
  return `.${variantName}${escapeClassName(separator)}${className.slice(1)}`
}
