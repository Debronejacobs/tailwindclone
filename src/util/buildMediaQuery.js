import _ from 'lodash'

function isInvalidValue(value) {
  return !value || value === '' || value === 0 || value === '0'
}

export default function buildMediaQuery(screens) {
  if (isInvalidValue(screens)) {
    return undefined
  }

  if (_.isString(screens)) {
    screens = { min: screens }
  }

  if (!Array.isArray(screens)) {
    screens = [screens]
  }

  return _(screens)
    .map((screen) => {
      if (_.has(screen, 'raw')) {
        return screen.raw
      }

      return _(screen)
        .map((value, feature) => {
          feature = _.get(
            {
              min: 'min-width',
              max: 'max-width',
            },
            feature,
            feature
          )

          if (isInvalidValue(value)) {
            return undefined
          }

          return `(${feature}: ${value})`
        })
        .uniq()
        .filter((v) => v !== undefined)
        .join(' and ')
    })
    .join(', ')
}
