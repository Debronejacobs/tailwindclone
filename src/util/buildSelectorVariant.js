import parser from 'postcss-selector-parser'
import tap from 'lodash/tap'

export default (selector, variantName, separator, onError = () => {}) =>
  parser(selectors => {
    tap(selectors.first.filter(({ type }) => type === 'class').pop(), classSelector => {
      if (classSelector === undefined) {
        onError('Variant cannot be generated because selector contains no classes.')
        return
      }

      classSelector.value = `${variantName}${separator}${classSelector.value}`
    })
  }).processSync(selector)
