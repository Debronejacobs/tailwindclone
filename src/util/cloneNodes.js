import _ from 'lodash'

export default nodes => _.map(nodes, node => node.clone())
