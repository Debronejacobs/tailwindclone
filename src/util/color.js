import namedColors from 'color-name'

let HEX = /^#([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})?$/i
let SHORT_HEX = /^#([a-f\d])([a-f\d])([a-f\d])([a-f\d])?$/i
let VALUE = `(?:\\d+|\\d*\\.\\d+)%?`
let SEP = `(?:,\\s*|\\s+)`
let ALPHA_SEP = `\\s*[,/]\\s*`
let RGB_HSL = new RegExp(
  `^(rgb|hsl)a?\\(\\s*(${VALUE})${SEP}(${VALUE})${SEP}(${VALUE})(?:${ALPHA_SEP}(${VALUE}))?\\s*\\)$`
)

export function parseColor(value) {
  if (typeof value !== 'string') {
    return null
  }

  value = value.trim()

  if (value in namedColors) {
    return {
      mode: 'rgb',
      color: namedColors[value],
    }
  }

  let hex = value
    .replace(SHORT_HEX, (_, r, g, b, a) => ['#', r, r, g, g, b, b, a ? a + a : ''].join(''))
    .match(HEX)

  if (hex !== null) {
    return {
      mode: 'rgb',
      color: [parseInt(hex[1], 16), parseInt(hex[2], 16), parseInt(hex[3], 16)],
      alpha: hex[4] ? parseInt(hex[4], 16) / 255 : undefined,
    }
  }

  let match = value.match(RGB_HSL)

  if (match !== null) {
    return {
      mode: match[1],
      color: [match[2], match[3], match[4]],
      alpha: match[5],
    }
  }

  return null
}

export function formatColor({ mode, color, alpha }) {
  let hasAlpha = alpha !== undefined
  return `${mode}${hasAlpha ? 'a' : ''}(${color.join(', ')}${hasAlpha ? `, ${alpha}` : ''})`
}
