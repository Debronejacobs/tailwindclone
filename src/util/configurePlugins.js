export default (pluginConfig, plugins) => {
  const pluginNames = Array.isArray(pluginConfig)
    ? pluginConfig
    : Object.keys(plugins).filter(
        pluginName => pluginConfig !== false && pluginConfig[pluginName] !== false
      )

  return pluginNames.map(pluginName => plugins[pluginName]())
}
