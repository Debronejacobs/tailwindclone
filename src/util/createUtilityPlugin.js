import _ from 'lodash'
import fromPairs from 'lodash/fromPairs'
import toPairs from 'lodash/toPairs'
import castArray from 'lodash/castArray'

function className(classPrefix, key) {
  if (key === '') {
    return classPrefix
  }

  if (key.startsWith('-')) {
    return `-${classPrefix}${key}`
  }

  return `${classPrefix}-${key}`
}

export default function createUtilityPlugin(themeKey, utilityVariations, mapTheme = theme => theme) {
  return function({ e, addUtilities, variants, theme }) {
    const utilities = utilityVariations.map(([classPrefix, properties]) => {
      return fromPairs(
        toPairs(mapTheme(theme(themeKey))).map(([key, value]) => {
          return [
            `.${e(className(classPrefix, key))}`,
            fromPairs(castArray(properties).map(property => [property, value])),
          ]
        })
      )
    })
    return addUtilities(utilities, variants(themeKey))
  }
}

export function convertDefaultKey(theme) {
  return _.mapKeys(theme, (value, key) => key === 'default' ? '' : key)
}
