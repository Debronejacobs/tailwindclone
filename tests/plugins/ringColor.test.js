import { quickPluginTest } from '../util/run'

quickPluginTest('ringColor', {
  safelist: [
    // Arbitrary values
    'ring-[#0088cc]',
    'ring-[var(--my-value)]',
  ],
  theme: {
    colors: {
      transparent: 'transparent',
      black: '#000',
      red: {
        100: '#f00',
        200: '#f00',
      },
    },
  },
}).toMatchSnapshot()

// With `ringOpacity` plugin enabled
quickPluginTest('ringColor', {
  safelist: [
    // Arbitrary values
    'ring-[#0088cc]',
    'ring-[var(--my-value)]',
  ],
  corePlugins: ['ringOpacity'],
  theme: {
    colors: {
      transparent: 'transparent',
      black: '#000',
      red: {
        100: '#f00',
        200: '#f00',
      },
    },
    opacity: {
      0: '0',
      50: '.5',
    },
  },
}).toMatchSnapshot()
