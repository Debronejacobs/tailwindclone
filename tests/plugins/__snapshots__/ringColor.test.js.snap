// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test the 'ringColor' plugin 1`] = `
"
.ring-\\[\\#0088cc\\] {
  --tw-ring-color: #08c;
}

.ring-\\[var\\(--my-value\\)\\] {
  --tw-ring-color: var(--my-value);
}

.ring-black {
  --tw-ring-color: #000;
}

.ring-black\\/0 {
  --tw-ring-color: #0000;
}

.ring-black\\/10 {
  --tw-ring-color: #0000001a;
}

.ring-black\\/100 {
  --tw-ring-color: #000;
}

.ring-black\\/15 {
  --tw-ring-color: #00000026;
}

.ring-black\\/20 {
  --tw-ring-color: #0003;
}

.ring-black\\/25 {
  --tw-ring-color: #00000040;
}

.ring-black\\/30 {
  --tw-ring-color: #0000004d;
}

.ring-black\\/35 {
  --tw-ring-color: #00000059;
}

.ring-black\\/40 {
  --tw-ring-color: #0006;
}

.ring-black\\/45 {
  --tw-ring-color: #00000073;
}

.ring-black\\/5 {
  --tw-ring-color: #0000000d;
}

.ring-black\\/50 {
  --tw-ring-color: #00000080;
}

.ring-black\\/55 {
  --tw-ring-color: #0000008c;
}

.ring-black\\/60 {
  --tw-ring-color: #0009;
}

.ring-black\\/65 {
  --tw-ring-color: #000000a6;
}

.ring-black\\/70 {
  --tw-ring-color: #000000b3;
}

.ring-black\\/75 {
  --tw-ring-color: #000000bf;
}

.ring-black\\/80 {
  --tw-ring-color: #000c;
}

.ring-black\\/85 {
  --tw-ring-color: #000000d9;
}

.ring-black\\/90 {
  --tw-ring-color: #000000e6;
}

.ring-black\\/95 {
  --tw-ring-color: #000000f2;
}

.ring-red-100 {
  --tw-ring-color: red;
}

.ring-red-100\\/0 {
  --tw-ring-color: #f000;
}

.ring-red-100\\/10 {
  --tw-ring-color: #ff00001a;
}

.ring-red-100\\/100 {
  --tw-ring-color: red;
}

.ring-red-100\\/15 {
  --tw-ring-color: #ff000026;
}

.ring-red-100\\/20 {
  --tw-ring-color: #f003;
}

.ring-red-100\\/25 {
  --tw-ring-color: #ff000040;
}

.ring-red-100\\/30 {
  --tw-ring-color: #ff00004d;
}

.ring-red-100\\/35 {
  --tw-ring-color: #ff000059;
}

.ring-red-100\\/40 {
  --tw-ring-color: #f006;
}

.ring-red-100\\/45 {
  --tw-ring-color: #ff000073;
}

.ring-red-100\\/5 {
  --tw-ring-color: #ff00000d;
}

.ring-red-100\\/50 {
  --tw-ring-color: #ff000080;
}

.ring-red-100\\/55 {
  --tw-ring-color: #ff00008c;
}

.ring-red-100\\/60 {
  --tw-ring-color: #f009;
}

.ring-red-100\\/65 {
  --tw-ring-color: #ff0000a6;
}

.ring-red-100\\/70 {
  --tw-ring-color: #ff0000b3;
}

.ring-red-100\\/75 {
  --tw-ring-color: #ff0000bf;
}

.ring-red-100\\/80 {
  --tw-ring-color: #f00c;
}

.ring-red-100\\/85 {
  --tw-ring-color: #ff0000d9;
}

.ring-red-100\\/90 {
  --tw-ring-color: #ff0000e6;
}

.ring-red-100\\/95 {
  --tw-ring-color: #ff0000f2;
}

.ring-red-200 {
  --tw-ring-color: red;
}

.ring-red-200\\/0 {
  --tw-ring-color: #f000;
}

.ring-red-200\\/10 {
  --tw-ring-color: #ff00001a;
}

.ring-red-200\\/100 {
  --tw-ring-color: red;
}

.ring-red-200\\/15 {
  --tw-ring-color: #ff000026;
}

.ring-red-200\\/20 {
  --tw-ring-color: #f003;
}

.ring-red-200\\/25 {
  --tw-ring-color: #ff000040;
}

.ring-red-200\\/30 {
  --tw-ring-color: #ff00004d;
}

.ring-red-200\\/35 {
  --tw-ring-color: #ff000059;
}

.ring-red-200\\/40 {
  --tw-ring-color: #f006;
}

.ring-red-200\\/45 {
  --tw-ring-color: #ff000073;
}

.ring-red-200\\/5 {
  --tw-ring-color: #ff00000d;
}

.ring-red-200\\/50 {
  --tw-ring-color: #ff000080;
}

.ring-red-200\\/55 {
  --tw-ring-color: #ff00008c;
}

.ring-red-200\\/60 {
  --tw-ring-color: #f009;
}

.ring-red-200\\/65 {
  --tw-ring-color: #ff0000a6;
}

.ring-red-200\\/70 {
  --tw-ring-color: #ff0000b3;
}

.ring-red-200\\/75 {
  --tw-ring-color: #ff0000bf;
}

.ring-red-200\\/80 {
  --tw-ring-color: #f00c;
}

.ring-red-200\\/85 {
  --tw-ring-color: #ff0000d9;
}

.ring-red-200\\/90 {
  --tw-ring-color: #ff0000e6;
}

.ring-red-200\\/95 {
  --tw-ring-color: #ff0000f2;
}

.ring-transparent {
  --tw-ring-color: transparent;
}

.ring-transparent\\/0 {
  --tw-ring-color: #0000;
}

.ring-transparent\\/10 {
  --tw-ring-color: #0000001a;
}

.ring-transparent\\/100 {
  --tw-ring-color: #000;
}

.ring-transparent\\/15 {
  --tw-ring-color: #00000026;
}

.ring-transparent\\/20 {
  --tw-ring-color: #0003;
}

.ring-transparent\\/25 {
  --tw-ring-color: #00000040;
}

.ring-transparent\\/30 {
  --tw-ring-color: #0000004d;
}

.ring-transparent\\/35 {
  --tw-ring-color: #00000059;
}

.ring-transparent\\/40 {
  --tw-ring-color: #0006;
}

.ring-transparent\\/45 {
  --tw-ring-color: #00000073;
}

.ring-transparent\\/5 {
  --tw-ring-color: #0000000d;
}

.ring-transparent\\/50 {
  --tw-ring-color: #00000080;
}

.ring-transparent\\/55 {
  --tw-ring-color: #0000008c;
}

.ring-transparent\\/60 {
  --tw-ring-color: #0009;
}

.ring-transparent\\/65 {
  --tw-ring-color: #000000a6;
}

.ring-transparent\\/70 {
  --tw-ring-color: #000000b3;
}

.ring-transparent\\/75 {
  --tw-ring-color: #000000bf;
}

.ring-transparent\\/80 {
  --tw-ring-color: #000c;
}

.ring-transparent\\/85 {
  --tw-ring-color: #000000d9;
}

.ring-transparent\\/90 {
  --tw-ring-color: #000000e6;
}

.ring-transparent\\/95 {
  --tw-ring-color: #000000f2;
}
"
`;

exports[`should test the 'ringColor' plugin 2`] = `
"
.ring-\\[\\#0088cc\\] {
  --tw-ring-opacity: 1;
  --tw-ring-color: rgb(0 136 204 / var(--tw-ring-opacity));
}

.ring-\\[var\\(--my-value\\)\\] {
  --tw-ring-color: var(--my-value);
}

.ring-black {
  --tw-ring-opacity: 1;
  --tw-ring-color: rgb(0 0 0 / var(--tw-ring-opacity));
}

.ring-black\\/0 {
  --tw-ring-color: #0000;
}

.ring-black\\/50 {
  --tw-ring-color: #00000080;
}

.ring-red-100 {
  --tw-ring-opacity: 1;
  --tw-ring-color: rgb(255 0 0 / var(--tw-ring-opacity));
}

.ring-red-100\\/0 {
  --tw-ring-color: #f000;
}

.ring-red-100\\/50 {
  --tw-ring-color: #ff000080;
}

.ring-red-200 {
  --tw-ring-opacity: 1;
  --tw-ring-color: rgb(255 0 0 / var(--tw-ring-opacity));
}

.ring-red-200\\/0 {
  --tw-ring-color: #f000;
}

.ring-red-200\\/50 {
  --tw-ring-color: #ff000080;
}

.ring-transparent {
  --tw-ring-color: transparent;
}

.ring-transparent\\/0 {
  --tw-ring-color: #0000;
}

.ring-transparent\\/50 {
  --tw-ring-color: #00000080;
}

.ring-opacity-0 {
  --tw-ring-opacity: 0;
}

.ring-opacity-50 {
  --tw-ring-opacity: .5;
}
"
`;
